_posts/2018-05-25-ecs-calculator.md

```diff
@@ -0,0 +1,138 @@
+|c4.2xlarge| 8192| 15040|
+|c4.4xlarge| 16384| 30145|
+|c4.8xlarge| 36864| 60385|
+|----------------------|
+|c3.large| 2048| 3765|
+|c3.xlarge| 4096| 7480|
+|c3.2xlarge| 8192| 15040|
+|c3.4xlarge| 16384| 30145|
+|c3.8xlarge| 32768| 60385|
+|----------------------|
+|r4.large| 2048| 15292|
+|r4.xlarge| 4096| 30664|
+|r4.2xlarge| 8192| 61408|
+|r4.4xlarge| 16384| 122881|
+|r4.8xlarge| 32768| 245857|
+|r4.16xlarge| 65536| 491809|
+|----------------------|
+|r3.large| 2048| 15299|
+|r3.xlarge| 4096| 30680|
+|r3.2xlarge| 8192| 61442|
+|r3.4xlarge| 16384| 122950|
+|r3.8xlarge| 32768| 245996|
+|----------------------|
+|i2.xlarge| 4096| 30680|
+|i2.2xlarge| 8192| 61442|
+|i2.4xlarge| 16384| 122950|
+|i2.8xlarge| 32768| 245996|
+|----------------------|
+|i3.large| 2048| 15292|
+|i3.xlarge| 4096| 30664|
+|i3.2xlarge| 8192| 61408|
+|i3.4xlarge| 16384| 122881|
+|i3.8xlarge| 32768| 245857|
+|i3.16xlarge| 65536| 491809|
+|----------------------|
+|m5.large| 2048| 7690|
+|m5.xlarge| 4096| 15586|
+|m5.2xlarge| 8192| 31377|
+|m5.4xlarge| 16384| 62960|
+|m5.12xlarge| 49152| 189294|
+|m5.24xlarge| 98304| 378665|
+|----------------------|
+|c5.large| 2048| 3714|
+|c5.xlarge| 4096| 7634|
+|c5.2xlarge| 8192| 15473|
+|c5.4xlarge| 16384| 31152|
+|c5.9xlarge| 36864| 70351|
+|c5.18xlarge| 73728| 140780|
+|----------------------|
+|d2.xlarge| 4096| 30664|
+|d2.2xlarge| 8192| 61408|
+|d2.4xlarge| 16384| 122881|
+|d2.8xlarge| 36864| 245857|
+
+</div>
+
+❗️You can also find this information in yaml format [here](/assets/files/2018-05-25-ecs-calculator/ec2-instance-list.yml) 👈
+
+The solution is to make the optimal choice of the EC2 instance type with its budgeted CPU and Memory units to efficiently fit ECS containers in the instance based on the data from the table. Therefore, it is important to first have some insight into the limits of CPU and Memory reservation.
+
+The CPU units can be reserved with soft limit because containers can burst above their provision. One container can burst above its allocated units if no other containers are taking up resources. Besides, containers also share their unallocated CPU units with other containers on the instance with the same ratio as their allocated amount.
+
+It is also useful to understand how your container can effectively use the memory resources.The ECS memory resource model is more complex then the CPU resource model and allows you to configure both memory reservations and memory limits. You can see the whole concept below:
+
+![memory](/assets/posts/2018-05-25-ecs-calculator/memory_resources.png)
+
+Let's assume your container normally uses 128 MBs of memory, but occasionally bursts to 256 MBs of memory for short periods of time. As an option, you can set a Memory Reservation of 128 MBs, and a Memory Limit of 300 MBs. This configuration would allow the container to only reserve 128 MBs of memory from the resources on the container instance, but also allows the container to consume more memory resources when needed.
```

Comment:
Let's assume your container normally uses 128 MBs of memory, but occasionally bursts to 256 MBs of memory for short periods of time. As an option, you can set a Memory Reservation of 128 MBs, and a Memory Limit of 300 MBs. This configuration would allow the container to only reserve 128 MBs of memory from the resources on the container instance, but also allows the container to consume more memory resources when needed. Another option is to only use the hard limit for the memory assignment. This way you can bound its container to a specific value and avoid containers fighting for resources and possibly crashing because of lack of memory units. Hard memory bound requires extensive profiling of your containers for optimal resource allocation.